endpoints/events/event.go

```diff
@@ -0,0 +1,363 @@
+	Bidder    string    `json:"bidder,omitempty"`
+	Timestamp int64     `json:"timestamp,omitempty"`
+}
+
+type TrackingPixel struct {
+	Content     []byte `json:"content,omitempty"`
+	ContentType string `json:"content_type,omitempty"`
+}
+
+type eventEndpoint struct {
+	Accounts      stored_requests.AccountFetcher
+	Analytics     analytics.PBSAnalyticsModule
+	Cfg           *config.Configuration
+	TrackingPixel *TrackingPixel
+}
+
+func NewEventEndpoint(cfg *config.Configuration, accounts stored_requests.AccountFetcher, analytics analytics.PBSAnalyticsModule) httprouter.Handle {
+	ee := &eventEndpoint{
+		Accounts:      accounts,
+		Analytics:     analytics,
+		Cfg:           cfg,
+		TrackingPixel: trackingPixelPng,
+	}
+
+	return ee.Handle
+}
+
+func (e *eventEndpoint) Handle(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	// parse event request from http req
+	eventRequest, err := ParseEventRequest(r)
+
+	// handle possible parsing errors
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(fmt.Sprintf("invalid request: %s\n", err.Error())))
+
+		return
+	}
+
+	// validate account id
+	accountId, err := validateRequiredParameter(r, AccountIdParameter)
+
+	if err != nil {
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte(fmt.Sprintf("Account '%s' is required query parameter and can't be empty", AccountIdParameter)))
+		return
+	}
+	eventRequest.AccountID = accountId
+
+	if eventRequest.Analytics != Enabled {
+		return
+	}
+
+	ctx := context.Background()
+
+	// get account details
+	account, errs := GetAccount(ctx, e.Cfg, e.Accounts, eventRequest.AccountID)
+	if len(errs) > 0 {
+		w.WriteHeader(http.StatusInternalServerError)
+		for _, err := range errs {
+			w.Write([]byte(fmt.Sprintf("Internal Error: %s\n", err.Error())))
+		}
+
+		return
+	}
+
+	// account does not support events
+	if !account.EventsEnabled {
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write([]byte(fmt.Sprintf("Account '%s' doesn't support events", eventRequest.AccountID)))
+		return
+	}
+
+	// handle notification event
+	e.Analytics.LogNotificationEventObject(&analytics.NotificationEvent{
+		Type:      string(eventRequest.Type),
+		Timestamp: time.Unix(eventRequest.Timestamp, 0),
+		Bidid:     eventRequest.Bidid,
+		Bidder:    eventRequest.Bidder,
+		Account:   account,
+	})
+
+	// OK
+	w.WriteHeader(http.StatusOK)
+	// Add tracking pixel if format == image
+	if eventRequest.Format == Image {
+		w.Header().Add("Content-Type", e.TrackingPixel.ContentType)
+		w.Write(e.TrackingPixel.Content)
+	}
+}
+
+/**
+ * Converts an EventRequest to an URL
+ */
+func EventRequestToUrl(externalUrl string, request *EventRequest) string {
+	s := fmt.Sprintf(TemplateUrl, externalUrl, request.Type, request.Bidid, request.AccountID)
+
+	return s + optionalParameters(request)
+}
+
+/**
+ * Parses an EventRequest from an Http request
+ */
+func ParseEventRequest(r *http.Request) (*EventRequest, error) {
```

Comment:
This function returns only the first error. In other places PBS tries to return everything that went wrong.
Do we want to do that here as well ?

Reply:
That came to mind as well since I saw that pattern in other places but, that would imply an intent on doing something with all of the errors - From a user's perspective, being aware of all the syntax errors with a single call would (probably) prevent future calls using the same payload.

However, PBS Java does a sequential validation of the request returning on the first error that it encounters:
https://github.com/rubicon-project/prebid-server-java/blob/master/src/main/java/org/prebid/server/handler/NotificationEventHandler.java#L69

As such, I opted for following that same approach and keep their responses similar.
If that is unnecessary I'm happy to change it.